### ISSUE TITLE
Training a Music Transformer Model Using REMI Representation

### ISSUE THREAD

**Issue by Fansesi** [Natooz/MidiTok#19]
Hello,

I’m trying to train a Music Transformer (paper: [https://arxiv.org/pdf/1809.04281.pdf](https://arxiv.org/pdf/1809.04281.pdf), implementation taken from: [https://colab.research.google.com/github/spectraldoy/MusicTransformerTensorFlow/blob/main/Music_Transformer_Public.ipynb](https://colab.research.google.com/github/spectraldoy/MusicTransformerTensorFlow/blob/main/Music_Transformer_Public.ipynb)) model with the REMI representation on Colab’s Tesla T4 GPU. I have trained a few models. While training their training loss was not decreasing as I expected but it’s my first time training a transformer model so I didn’t think about it too much. When I did the inference though, trained model just gave me 1’s which is my `SOS` token (numbers other than 1’s are inputs, except 67): 

```
idx_list parameter is: [154 221  17  68  85 158 221  20  68  85 162 221  25  68  85 166 221  29
  68  85 170 221  13  68  97  32  68  93 174 221  20  68  85   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1  67   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1]
```

I thought it had been overfitted or underfitted. Because my loss was relatively high (around 3.97) I suspected it has underfitted. But even if I train it quite long enough I'm still getting the 1's. So I’ve started searching for the source of the problem. Now I have couple of things in my mind: 
First of all, I’m aware that training even longer might potentially solve the problem. But I'm guessing that training not long enough shouldn’t end up with just 1’s. 
As a second thing, I’ve suspected my learning rate. I’m using a custom learning schedule and it’s graph is like this:

![index](https://user-images.githubusercontent.com/101516158/204153785-a44db601-c450-4141-bd31-065f399e4099.png)

But when I try to train the model with a static learning rate like 0.1 or 0.01, my loss isn’t decreasing enough either. And I’m still getting the 1’s in the inference phase.  

And at last, the main problem I think: in the data processing, after I convert my .mid files into tokens using MidiTok, I truncate and pad the sequences according to a `seq_len` parameter which I set before. For example, let’s say `seq_len=256` . If a .mid files token length is bigger than 256 I first truncate it as many as I can, then pad the remaning. So if the .mid files token length is 1000, I first cut 3 pieces of 256 token long excerpts and pad the remaining 1000-256.3 = 232 tokens to 256. And if the .mid files token length is lower than the `seq_len` I just pad it. For example .mid files token length is 250, I pad it to 256.

But now to think of it, this method makes no sense because I’m changing the order of the tokens which basicly means that I'm feeding the model some meaningless integers. I’m thinking two approaches on processing the tokens and equalizing their lengths right now but I simply cannot decide what to do. Here are the two approaches that came to my mind:

1) I’ll find a suitable (by suitable I mean, I’ll find an index of a `bar` token which is lower than the `seq_len`) `bar` token and truncate it from there and pad it to `seq_len` afterwards. And just before training I’ll add `sos` and `eos` tokens to inputs first index and targets last index respectively. With this way the model will always learn from the beginning of a bar. 

2) My other approach is to increase the `seq_len` quite high and don’t truncate the tokens but just pad them if needed. This way I’ll be sure it’s going to start from the beginning of a song. But this way is problematic because if I increase the `seq_len`, I’ll need to decrease my batch size. And from what I’ve read on comparison papers, keeping the batch size high enough is generally good practice.

So I have a couple of questions:
* How do I need to equalize my tokens? What is the best practice for this?
* What should I expect from my training loss? I'm currently comparing it with the [https://colab.research.google.com/github/spectraldoy/MusicTransformerTensorFlow/blob/main/Music_Transformer_Public.ipynb](https://colab.research.google.com/github/spectraldoy/MusicTransformerTensorFlow/blob/main/Music_Transformer_Public.ipynb)'s training loss. Researcher had achieved a loss of 1.257 with approx. 100.000 training steps and I don't think I can get a loss near that even if I train the model for 100.000 steps.
* And I am not sure of my `seq_len` parameter. How do we specify that parameter and batch size? Just by experimenting or are there any rules for this?
* Generally speaking, what could be the problem if the preprocessing that I'm doing right now is correct? Where should I look for the problem?

Sorry for such a long question but I felt like I should explain everything. My mind is quite messed up and I don't have time or resources to try out all the ideas that came to my mind. So all advices are welcome.

Thanks in advance!

Fansesi

**Comment by Natooz** [comment]
Hi, 👋
I see there is a lot of concerns, as you said time is limited so you did good to ask ! 😄

I'll first answer to the questions bullet by bullet:
* I by equalize you mean truncate the tokens according to a `seq_len`parameter, you did good ! What I personally do is to load the tokens through a PyTorch [Dataset](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html), slice them in sequences between a min and max length, and then pad them altogether. The code is below.
* The final loss value to expect will depend on several factors: the size and type of model, the loss function, the data (is it easy / hard to learn from) and the vocabulary size. For the Remi tokenization, a Transformer of 8 layers of dim 512 and CrossEntropy loss you should expect to rapidly reach a loss value close to 2.
* There is no general rule regarding `seq_len`, except that it should be enough to cover a range of tokens used during inference. In some conditions you might want `seq_len` to be very high, eg story generation where the context is relatively long. You can measure the average equivalent of 256 tokens in beats if you want. You can also increase the batch size by doing gradient accumulation if you think it isn't high enough, by I would no bother, 16 should be enough.
* I think the problem is either the training / target sequence that is wrong, or the inference code that does not generate properly. During training the target sequence should be the same than the input sequence but shifted by one, as we expect the model to predict the next token at each position/index (method called teacher forcing). Now the model should never predict `SOS` tokens, as it should never have been trained with a sequence containing this token. So either the training is not right, either the inference algorithm does not expand auto regressively the input sequence at each step.

By the way I took a quick look at the notebook, I suggest you to use top-k or nucleus sampling instead of greedy decoding.
https://openreview.net/forum?id=rygGQyrFvH
https://huggingface.co/blog/how-to-generate
And the [GPT2 implementation of HuggingFace](https://huggingface.co/docs/transformers/model_doc/gpt2) is very convenient to generate, as you can reuse all the calculations of past positions, which is much much faster.


The Dataset to load tokens:
```Python
from torch.nn.utils.rnn import pad_sequence
from torch.utils.data import Dataset

class MIDIDataset(Dataset):
    r"""Dataset for generator training

    :param data_path: path containing the real data to load, ex: 'data/death_metal_dataset'.
    :param min_seq_len: minimum sequence length (in nb of tokens)
    :param max_seq_len: maximum sequence length (in nb of tokens)
    :param padding_token: padding token, usually 0.
    :param tokenizer: tokenizer object, to use when fake_data_path is a list of MIDI paths. (default: None)
    """

    def __init__(self, data_path: Path, min_seq_len: int, max_seq_len: int, padding_token: int,
                 tokenizer: MIDITokenizer = None):
        samples = []
        as_midi = False
        files_paths = list(Path(data_path).glob(f'**/*.json'))
        if len(files_paths) == 0:
            files_paths = list(Path(data_path).glob(f'**/*.mid'))
            as_midi = True

        for file_path in tqdm(files_paths, desc=f'Preparing data {data_path.name}'):
            if as_midi:
                tokens = tokenizer.midi_to_tokens(MidiFile(file_path))[0]  # first track
            else:
                with open(file_path) as json_file:
                    tokens = json.load(json_file)['tokens'][0]  # first track
            i = 0
            while i < len(tokens):
                if i >= len(tokens) - min_seq_len:
                    break  # last sample is too short
                samples.append(LongTensor(tokens[i:i + max_seq_len]))
                i += len(samples[-1])  # could be replaced with max_seq_len

        self.samples = pad_sequence(samples, batch_first=True, padding_value=padding_token)

    def __getitem__(self, idx) -> Tuple[LongTensor, int]:
        return self.samples[idx]

    def __len__(self) -> int: return len(self.samples)

    def __repr__(self): return self.__str__()

    def __str__(self) -> str: return 'No data loaded' if len(self) == 0 else f'{len(self.samples)} samples'
```

Just before training, we shift by one the sequence.
```Python
def collate_gen(batch: List[LongTensor]) -> Tuple[LongTensor, LongTensor]:
    batch = stack(batch)  # (N,T)
    return batch[..., :-1], batch[..., 1:]  # input, target
```

**Comment by Fansesi** [comment]
Thanks for your response!

I had already aware of the teacher forcing method but in any case I have double checked my input and target datasets. They are all good and fine, `sos` and `eos` tokens are placed correctly but the problem is actually just a typo! I have a `train_step(inps, tars)` function but when I use it in training, it seems like I was giving my input dataset and target dataset as `train_step(tar_ds, inp_ds)`. I've fixed it now. 

I'll train the model a few thousand or more steps as soon as I can and try to do the inference using one of the methods you've described. When I'm done with these, I'll publish my updates here. 

As for the dataset processing pipeline, I'll try to use my main approach for now but if I can't get the results that I wanted I'll probably implement your codes in Tensorflow and use it. Thank you for the codes as well!

And I'll check out the GPT2 implementation when I have the time. 

Thank you again!

**Comment by Fansesi** [comment]
Hi again!

I have trained a model and I’m trying to do the inference phase now. I’ve looked into the GPT-2 implementation but I couldn’t see any code samples to use it for a custom model. It looks so high level and cannot be used for a model which is not inherited from the HuggingFace's modules. If there is a way to use it without changing my model pipeline, I would use it of course. 

So I decided to look for implementations in Tensorflow and I've written something like this:

```python
def top_p_logits(logits, p):
    with tf.compat.v1.variable_scope('top_p_logits'):
        logits_sort = tf.sort(logits, direction='DESCENDING')
        probs_sort = tf.nn.softmax(logits_sort)
        probs_sums = tf.cumsum(probs_sort, axis=1, exclusive=True)
        logits_masked = tf.compat.v1.where(probs_sums < p, logits_sort, tf.ones_like(
            logits_sort)*1000)  # [batch_size, vocab]
        min_logits = tf.reduce_min(input_tensor=logits_masked, axis=1, keepdims=True)  # [batchsize, 1]
        return tf.compat.v1.where(
            logits < min_logits,
            tf.ones_like(logits, dtype=logits.dtype) * -1e10,
            logits,
        )

def create_sample(model, temperature, top_p, length, first_inp=[[1]], EOS_token=2, stats=True):
  """Generates new tokens using top_p_logits function.
  @Params
  @model: transformer model to use.
  @temperature: temperature value
  @top_p: top_p value.  
  @first_inp: first input to the model. Should be of shape (1, inp_len). 
  @length: max length of the final sequence.
  """
  input=first_inp[0]
  output = first_inp 
 
  start = time.time()

  for i in range(0, length):
  # Get logits with a forward pass in our model
    logits = model(input)

    logits = logits[0][-1][:] / temperature #I think there is a problem with this line but I'm not sure what it is.

    filtered_logits = top_p_logits(logits, p=top_p)
    probabilities = tf.nn.softmax(filtered_logits, axis=-1)
    next_token = tf.random.categorical(probabilities, num_samples=1) 

    if next_token.numpy().tolist()[0][0]==EOS_token:
      print("Model predicted the EOS token!")
      break
    
    #defining the newly created token as the next input.
    output[0].append(next_token.numpy().tolist()[0][0])  
    input=next_token

  end = time.time()
  
  if stats:
    print(f"Time taken: {round(end - start, 2)} secs .")
    print(f"Generated total of {len(output[0])-len(first_inp[0])} tokens.")
    print(f"Generated list is: {output[0]}")

  return [output[0]] 
```

And for the generation, audiation and saving the generated midi:

```python
#define the model
transformer = TransformerDecoder(
    num_layers, d_model, num_heads, dff, vocab_size2, MAX_LENGTH, max_abs_position,
    use_bias, dropout_rate, layernorm_eps, tie_emb
) #These variables are defined before.

# build the model
_ = transformer(tf.random.uniform((1, MAX_LENGTH))) 

del _

# load the weights
transformer.load_weights(CONFIG.model_save_path + "model_30917_train_steps.h5") 

def midi_out(output_list, tokenizer, midi_name, save=False, output_path=""):
  midi_out = tokenizer.tokens_to_midi(output_list, [[24,False]])
  if save:
    midi_out.dump(output_path+f"{midi_name}.mid")
    print("Midi saved succesfully!")
  return midi_out 

max_decode_length=500
temperature = 1.0
top_p = 0.9

my_inp = [1] 

output_list = create_sample(
    model=transformer, 
    first_inp = [my_inp],
    temperature=temperature,
    top_p=top_p,
    length=max_decode_length, 
    )

output_path = CONFIG.output_dir #I'm taking my output_path from a config file.
midi_name = "generated_midi"
my_midi = midi_out(output_list, tokenizer=tokenizer, midi_name=midi_name, save=True, output_path=output_path) 
```

When I run the codes above, save the midi file and listen it, I only hear one or two notes. They are usually far apart from each other don't follow any pattern. I’m suspecting it is because of my model’s ability to generate but I am not sure of my inference code as well, so I decided to ask it here. 

And also I have some questions reagarding to topic. I’ll be glad if you can answer them:

- How does MidiTok handles tokens that don’t make sense? Is it just ignores them?
- Is it possible for a model to generate all of the tokens perfectly? By perfectly I mean, can a model create tokens that all of them make sense regarding to token placement just like in a tokenized song? I’m not talking about it’s musicality though. Even if the generation doesn't sound good, is it possible that all of the generated tokens can mean something?
- Should I feed all of the data (in time step *i*, inputs and all the generated tokens at time steps *1,2,3...(i-1)*) every time while generating the next token or only the previously generated token? If the answer is just the previously generated token, how does the model generate next tokens according to all of the previously generated tokens? (I think my generation code may be wrong because of this dilemma)
- What do you think about the source of the problem: is my sampling code wrong or is it that my model isn’t good enough or something else?

And again, sorry for the trouble I'm making but my deadline is getting closer and I should probably start writing my report 😃. 

Thanks in advance!

Fansesi

**Comment by Natooz** [comment]
Hi 👋,

I had some time this afternoon, I just uploaded [a notebook using HF GPT2](https://github.com/Natooz/MidiTok/blob/main/colab-notebooks/GPT2_Transformer_Workflow.ipynb).
I did not test / run it, there might be some environment / path bugs but the code blocks are working.
You can see how to create the model, train it (with an external function) and use it afterwards. The generate method reuses previous attention computations so we don't have to reuse them again.

For other questions:

* MidiTok will just ignore them yes, so that if a model predicts some errors these won't crash / discard the whole token sequence;
* Yes it can, if it has been well trained. Tokenizers have a [’token_types_errors’](https://github.com/Natooz/MidiTok/blob/main/miditok/remi.py#L279) method that can measure the rate of errors of predictions. In practice, a mid-size transformer (8 layers, dim 512) should yield very few errors (<1%);
* For an input sequence $\mathbf{x}$, as we use a causal attention mask we will have $p\_\theta (\mathbf{x})\_i \propto \mathbf{x}\_{\leq i}$ ( $p\_{\theta}$ is the model), i.e. the hidden state at position i will only depend on the input tokens at positions <= to i. So when generating the next token, you still need to compute it depending on the past ones. But as past tokens computations do not depend on future ones, it means we don't need to recompute them, and that is what GPT2 HF implementation does, and it's faster. You just need to find the good balance between sequence size, and the number of past computations to store (can rapidly use GPU memory);
* `logits = logits[0][-1][:] / temperature` : this should be ok if the dimensions correspond to `batch, sequence, logits`, but I think `top_p_logits()` takes inputs tensors with dimensions `batch, logits`. If that's the case just replace with `logits = logits[:, -1] / temperature`. For the rest I am not a TF user to I cannot tell if something if wrong with sampling, I suggest you to use a good debugger and check dimensions and values. The problem could also come from the model not being well trained or not big enough.

No worry, I'm always happy to help ! 😄

**Comment by Fansesi** [comment]
Hi, 
Sorry for replying so late, I've been busy with finishing the project 😄. I have succesfully done the inference but with `top_k` decoding algorithm rather than `top_p` decoding algorithm. I guess there is a problem with my code but I can't find where it is. It's fine for me to switch to `top_k` so I don't mind. 

When I finish creating my Github repo, I can publish it here as well. Maybe it could help someone in their way of training a Music Transformer 😄.

And thanks again @Natooz for helping me out. I couldn't have done it without you!

Fansesi
